src/pymor/reductors/loewner.py

```diff
@@ -0,0 +1,255 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.core.base import BasicObject
+from pymor.models.iosys import LTIModel
+from pymor.models.transfer_function import TransferFunction
+from pymor.tools.random import new_rng
+
+
+class LoewnerReductor(BasicObject):
+    """Reductor based on Loewner interpolation framework.
+
+    The reductor implements interpolation based on the Loewner framework as in :cite:`ALI17`.
+
+    Parameters
+    ----------
+    s
+        |Numpy Array| of shape (n,) containing the frequencies.
+    Hs
+        |Numpy Array| of shape (n, p, m) or |TransferFunction| resembling the transfer function
+        samples.
+    partitioning
+        `str` or `tuple` of length 2. Strings can either be 'even-odd' or 'half-half' defining
+        the splitting rule. A user-defined partitioning can be defined by passing a tuple of the
+        left and right indices. Defaults to `even-odd`.
+    ordering
+        The ordering with respect to which the splitting rule is executed. Can be either
+        'magnitude', 'random' or 'regular'. Defaults to 'regular'.
+    ltd
+        |Numpy Array| representing left tangential directions. If `None` tangential directions will
+        be chosen randomly (normal distribution).
+    rtd
+        |Numpy Array| representing right tangential directions. If `None` tangential directions will
+        be chosen randomly (normal distribution).
+    """
+
+    def __init__(self, s, Hs, partitioning='even-odd', ordering='regular', ltd=None, rtd=None):
+        self.__auto_init(locals())
+
+    def reduce(self, tol=1e-7):
+        """Reduce using Loewner framework.
+
+        Parameters
+        ----------
+        tol
+            Truncation tolerance for rank of Loewner matrices.
+
+        Returns
+        -------
+        rom
+            Reduced |LTIModel|.
+        """
+        L, Ls, V, W = loewner_quadruple(self.s, self.Hs, partitioning='even-odd', ordering='regular',
+                                        ltd=None, rtd=None)
+        LLS = np.vstack([L, Ls])
+        Y, S, _ = spla.svd(LLS, full_matrices=False)
```

Comment:
The SVDs should be cached. Especially in the current implementation where the partitioning and data does not change.  It might also be nice to pass the `partitioning` parameter to the `reduce` method, but then a more sophisticated caching mechanism would be needed.

Reply:
I added caching just now. I will have to think about passing the `partitioning` to  `reduce`. It makes sense, but would also mean having some caching mechanism that maps partitionings to corresponding SVDs... Maybe it's easier to leave it as is.